crates/admin/src/cluster_controller/nodeset_selection.rs

```diff
@@ -0,0 +1,541 @@
+        assert!(
+            min_copies < u8::MAX >> 1,
+            "The replication factor implies a cluster size that exceeds the maximum supported size"
+        );
+        let optimal_fault_tolerant_nodeset_size = (min_copies as usize - 1) * 2 + 1;
+        assert!(
+            optimal_fault_tolerant_nodeset_size >= min_copies as usize,
+            "The calculated minimum nodeset size can not be less than the replication factor"
+        );
+
+        let (nodeset_min_size, nodeset_target_size) = match strategy {
+            NodeSetSelectionStrategy::FaultTolerantAdaptive => {
+                (min_copies as usize, optimal_fault_tolerant_nodeset_size)
+            }
+        };
+
+        let nodeset = match strategy {
+            NodeSetSelectionStrategy::FaultTolerantAdaptive => {
+                let mut nodes = preferred_nodes
+                    .iter()
+                    .copied()
+                    .filter(|node_id| candidates.0.contains(node_id))
+                    .choose_multiple(rng, nodeset_target_size);
+
+                if nodes.len() < nodeset_target_size {
+                    let remaining = nodeset_target_size - nodes.len();
+                    nodes.extend(
+                        candidates
+                            .into_iter()
+                            .filter(|node_id| !preferred_nodes.contains(node_id))
+                            .choose_multiple(rng, remaining),
+                    );
+                }
+
+                let nodes_len = nodes.len();
+                let nodeset = NodeSet::from_iter(nodes);
+                assert_eq!(
+                    nodeset.len(),
+                    nodes_len,
+                    "We have accidentally chosen duplicate candidates during nodeset selection"
+                );
+                nodeset
+            }
+        };
+
+        // todo: implement location scope-aware selection
+        if nodeset.len() < nodeset_min_size {
+            trace!(
+                selected_nodes_count = ?nodeset.len(),
+                ?nodeset_target_size,
+                "Failed to place replicated loglet: insufficient writeable nodes in the nodeset to meet availability goal"
+            );
+            return Err(NodeSelectionError::InsufficientWriteableNodes);
+        }
+
+        Ok(nodeset)
+    }
+}
+
+#[cfg(feature = "replicated-loglet")]
+#[derive(Debug, Clone, Default)]
+pub enum NodeSetSelectionStrategy {
+    /// Selects an optimal nodeset size based on the replication factor. The nodeset size is 2f+1,
+    /// working backwards from a replication factor of f+1.
+    ///
+    /// In this mode we aim to reach the optimal target size, but will settle for less as long as
+    /// the replication factor requirement is met. This strategy maximizes availability as long as
+    /// there are sufficient available storage nodes.
+    #[default]
+    FaultTolerantAdaptive,
```

Comment:
What about starting with a `AtLeastFaultTolerant`/`MaximumNodeSet` (pick a good name) strategy where we choose at least 2f + 1 nodes but if there are more nodes, then we take what we can get (being really greedy ;-)). That way we ensure that we don't run into the problem of risking sealability (when loosing f nodes) when returning a nodeset which is smaller than 2f + 1 and we maximize write availability.